mewarpx/mewarpx/assemblies.py

```diff
@@ -488,11 +490,17 @@ def isinside(self, X, Y, Z, aura=0):
             result (np.ndarray): array of flattened grid where all tiles
                 inside the cylinder are 1, and other tiles are 0.
         """
-        dist_to_center = np.sqrt(
+        dist_to_cyl_center_r = np.sqrt(
             (X - self.center_x)**2 + (Y - self.center_y)**2
         )
-        dr = self.r_outer + aura - self.r_center
-        result = np.where(abs(dist_to_center - self.r_center) <= dr, 1, 0)
+        dr = self.r_outer + 2.0*aura - self.r_inner
+        dz = self.zmax + 2.0*aura - self.zmin
+
+        result = np.where(
+            (abs(dist_to_cyl_center_r - self.r_center) <= dr/2.0)
+            & (abs(Z - (self.zmax + self.zmin)/2.0) <= dz/2.0),
+            1, 0
+        )
```

Comment:
I'd be happier with the readability if the code were written as the more simplistic 

```python
R >= r_inner - aura and R <= r_outer + aura and Z >= zmin - aura and Z <= zmax + aura
```
Is there a reason not to write it that way?

Also, I was thrown off by existing naming, in that `center_x` and `center_y` refer to the (currently origin-pinned) position of the cylinder, while `r_center` refers to the midpoint of the object's thickness.

Pedantically, this should really be named `AnnulusZ` rather than `CylinderZ`

Reply:
Oh, actually, this is Python, we can literally just do

```python
(r_inner - aura <= R <= r_outer + aura) and (zmin - aura <= Z <= z_max + aura)
```

Reply:
(I'm writing `R` where you have `dist_to_cyl_center_r`)

Reply:
Mostly just to minimize the number of operations performed on every array element (2 in the current case versus 4 in your way). I can certainly add a comment with the easier to read logic.
I agree in some use cases `AnnulusZ` would be more appropriate but in most use cases this assembly will be used for cylinders, in which case calling it an annulus would again be strange. I feel that a cylinder with defined inner and outer radii is a more sensible object than an annulus with 0 inner radius but I also know that it is a matter of personal opinion.

Reply:
Unfortunately, `(r_inner - aura <= R <= r_outer + aura) and (zmin - aura <= Z <= z_max + aura)` only works if `R` is a float, not when it is a numpy array.

Reply:
One thing that's unclear based on the doc comments - is this being called essentially once during initialization on the coordinates of the grid points, or is it being called on actual particle positions step after step?

If it's grid initialization, I think the clearer way is preferred. If it's particle positions, then we can think a little harder about whether even tighter expressions might be possible.

Reply:
Also, I'm curious - does the chained comparison actually produce multiple passes over array, or does numpy apply the chained comparison directly to each element?

Reply:
It is called on actual particle positions. I'll update the doc string accordingly.

Reply:
Ah, apparently chained comparisons are simply not allowed in NumPy